File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;

        sort(nums.begin(), nums.end());
        for (int i=0; i<nums.size(); ++i) {
            if (i > 0 && nums[i] == nums[i-1])
                continue;
                
            int l=i+1, r=nums.size()-1;
            while (l < r) {
                int tmp = nums[i] + nums[l] + nums[r];
                if (tmp == 0) {
                    ans.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    while (l < r && nums[l] == nums[l-1])
                        l++;
                    r--;
                    while (l < r && nums[r] == nums[r+1])
                        r--;
                }
                else if (tmp > 0) {
                    r--;
                }
                else {
                    l++;
                }
            }
        }

        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        
        ans = []
        for i in range(len(nums)):
            if i>0 and nums[i] == nums[i-1]:
                continue
                
            l, r = i+1, len(nums)-1
            while l < r:
                tmp = nums[i]+nums[l]+nums[r]
                if tmp==0:
                    ans.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l]==nums[l-1]:
                        l += 1
                    r -= 1
                    while l < r and nums[r]==nums[r+1]:
                        r -= 1
                elif tmp > 0:
                    r -= 1
                else:
                    l += 1

        
        return ans